src/daemon/distribution.c

```diff
@@ -0,0 +1,265 @@
+#include "distribution.h"
+#include <math.h>
+#include <pthread.h>
+
+typedef struct {
+  double max_boundary;
+  uint64_t counter;
+} bucket_t;
+
+struct distribution_s {
+  bucket_t *buckets;
+  double sum_gauges;
+  size_t num_buckets;
+  pthread_mutex_t mutex;
+};
+
+static bucket_t *bucket_new_linear(size_t num_buckets, double size) {
+  bucket_t *buckets = (bucket_t *)calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = (double)(i + 1) * size;
+  }
+
+  buckets[num_buckets - 1].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_exponential(size_t num_buckets, double initial_size,
+                                        double factor) {
+  bucket_t *buckets = (bucket_t *)calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  double multiplier = initial_size;
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = factor * multiplier;
+    multiplier *= initial_size;
+  }
+
+  buckets[num_buckets - 1].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_custom(size_t num_buckets,
+                                   const double *custom_buckets_sizes) {
+  bucket_t *buckets = (bucket_t *)calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  double ptr = 0.0;
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    ptr += custom_buckets_sizes[i];
```

Comment:
I think it would be more intuitive to pass bucket boundaries instead of bucket sizes. In other words, I think:

```c
double boundaries[] = {1, 2, 5, 10, 20, 50, 100};
```

is easier to read than:

```c
double sizes[] = {1, 1, 3, 5, 10, 30, 50};
```

Reply:
Thanks! Changed